### PR TITLE
[autoscaler] Split autoscaler interface public private

### LINKED ISSUE #10837: [autoscaler] Expose stable public API, move internals to _private package
Expose stable public API for scripts.py, tags.py, node_provider.py, move all other interfaces to private. We might have to also expose some more internal interfaces as needed.

A number of users are currently using unstable internal APIs of the autoscaler. Currently, the only stable public interface is the YAML file and node_provider. We should make this more explicit so platforms can build on the autoscaler with stability guarantees across Ray versions.

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This is an initial stab at splitting the autoscaler public/private interface. It moves most of autoscaler (verbatim) into _private, and exposes tags.py, node_provider.py, and a new sdk.py.

The public interface should be considered WIP at this point pending refinement. Since this PR is moving a lot of files, we should try to merge it quickly and iterate.

## Related issue number

Closes https://github.com/ray-project/ray/issues/10837

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/README.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+## Note on interface stability.
+
+All the public Python methods and attributes declared in this package can be considered stable public interfaces (except for those in the _private package).
```
Comment (richardliaw) [inline python/ray/autoscaler/README.rst:3]: nit - can you move this into the __init__ or sdk.py file?

**Review by richardliaw — APPROVED** [review]
This looks fine to me as treating this migration as a wip; maybe get an explicit approval from @alanwguo or @pcmoritz too.

**Comment by ericl** [comment]
@alanwguo @pcmoritz I'm merging this for now to avoid merge conflicts (since it moves every single file). Please let me know suggestions on follow-up changes to sdk.py. Btw, if you depend on autoscaler.* you can now of course still depend on autoscaler._private.*... as a temporary workaround, but long term it would be great to standardize on all users only depending on the public API.

**Comment by AmeerHajAli** [comment]
Can we make `_get_node_provider` public (without the underscore)?

**Comment by ericl** [comment]
No, why do you need to access that interface? That would expose all of node
provider to the public API. We can instead add more calls to sdk.py

**Pull request merged**